File: Lista01/PY/ex04.py
# Classe base Animal
class Animal:
    def __init__(self, nome):
        self.nome = nome

    # Método genérico de som (a ser sobrescrito nas subclasses)
    def som(self):
        pass

# Classe derivada Cachorro
class Cachorro(Animal):
    def som(self):
        print(f"{self.nome} faz: Au Au Au Au!")

# Classe derivada Gato
class Gato(Animal):
    def som(self):
        print(f"{self.nome} faz: Miauuuuuuuu!")

# Exemplo de uso das classes
cachorro = Cachorro("Pretinho")
gato = Gato("Olívio")

# Chamando o método som para cada objeto
cachorro.som()  # Saída: Rex faz: Au Au!
gato.som()      # Saída: Mingau faz: Miau!

File: Lista01/Java/ex04.java
public class ex04 {
    // Classe base Animal
    public static class Animal {
        void som() {
            System.out.println("O animal faz um som.");
        }
    }

    // Classe Cachorro que herda de Animal
    public static class Cachorro extends Animal {
        @Override
        void som() {
            System.out.println("O cachorro faz: Au Au!");
        }
    }

    // Classe Gato que herda de Animal
    public static class Gato extends Animal {
        @Override
        void som() {
            System.out.println("O gato faz: Miau!");
        }
    }

    public static void main(String[] args) {
        Animal cachorro = new Cachorro();
        Animal gato = new Gato();

        cachorro.som(); // O cachorro faz: Au Au!
        gato.som();     // O gato faz: Miau!
    }
}

